consul/acl_replication.go

```diff
@@ -0,0 +1,351 @@
+				ACL: *(remoteIter.Front()),
+			})
+			remoteIter.Next()
+			continue
+		}
+
+		// If the remote list is exhausted, then process this as a local
+		// delete. We know from the loop condition that there's something
+		// in the local list.
+		if remoteIter.Front() == nil {
+			changes = append(changes, &structs.ACLRequest{
+				Op:  structs.ACLDelete,
+				ACL: *(localIter.Front()),
+			})
+			localIter.Next()
+			continue
+		}
+
+		// At this point we know there's something at the front of each
+		// list we need to resolve.
+
+		// If the remote list has something local doesn't, we add it.
+		if localIter.Front().ID > remoteIter.Front().ID {
+			changes = append(changes, &structs.ACLRequest{
+				Op:  structs.ACLSet,
+				ACL: *(remoteIter.Front()),
+			})
+			remoteIter.Next()
+			continue
+		}
+
+		// If local has something remote doesn't, we delete it.
+		if localIter.Front().ID < remoteIter.Front().ID {
+			changes = append(changes, &structs.ACLRequest{
+				Op:  structs.ACLDelete,
+				ACL: *(localIter.Front()),
+			})
+			localIter.Next()
+			continue
+		}
+
+		// Local and remote have an ACL with the same ID, so we might
+		// need to compare them.
+		l, r := localIter.Front(), remoteIter.Front()
+		if r.RaftIndex.ModifyIndex > lastRemoteIndex && !r.IsSame(l) {
+			changes = append(changes, &structs.ACLRequest{
+				Op:  structs.ACLSet,
+				ACL: *r,
+			})
+		}
+		localIter.Next()
+		remoteIter.Next()
+	}
+	return changes
+}
+
+// FetchLocalACLs returns the ACLs in the local state store.
+func (s *Server) fetchLocalACLs() (structs.ACLs, error) {
+	_, local, err := s.fsm.State().ACLList()
+	if err != nil {
+		return nil, err
+	}
+	return local, nil
+}
+
+// FetchRemoteACLs is used to get the remote set of ACLs from the ACL
+// datacenter. The lastIndex parameter is a hint about which remote index we
+// have replicated to, so this is expected to block until something changes.
+func (s *Server) fetchRemoteACLs(lastRemoteIndex uint64) (*structs.IndexedACLs, error) {
+	args := structs.DCSpecificRequest{
+		Datacenter: s.config.ACLDatacenter,
+		QueryOptions: structs.QueryOptions{
+			Token:         s.config.ACLReplicationToken,
+			MinQueryIndex: lastRemoteIndex,
+			AllowStale:    true,
+		},
+	}
+	var remote structs.IndexedACLs
+	if err := s.RPC("ACL.List", &args, &remote); err != nil {
+		return nil, err
+	}
+	return &remote, nil
+}
+
+// UpdateLocalACLs is given a list of changes to apply in order to bring the
+// local ACLs in-line with the remote ACLs from the ACL datacenter.
+func (s *Server) updateLocalACLs(changes structs.ACLRequests) error {
+	var ops int
+	start := time.Now()
+	for _, change := range changes {
+		// Do a very simple rate limit algorithm where we check every N
+		// operations and wait out to the second before we continue. If
+		// it's going slower than that, the sleep time will be negative
+		// so we will just keep going without delay.
+		if ops >= s.config.ACLReplicationApplyLimit {
```

Comment:
I would change the smoothing to sleep `((1/ACLReplicationApplyLimit)-elapsed)` after every operation. That way it avoids bursty behavior which is more likely to cause IO contention